### ISSUE TITLE
log-driver=fluentd: docker refuses to start a container if fluentd is running

### ISSUE THREAD

**Issue by newhook** [docker/docker#15757]
I realize this is the documented behavior:

```
If container cannot connect to the Fluentd daemon, the container stops immediately.
```

However, this seems like a bad thing. Docker seems to correctly handle fluentd disappearing, dying and restarting so docker is clearly capable of handling these cases (it would be nice to document that, btw).

**Comment by dpgeekzero** [comment]
At the very least, falling back to json-file would be better than crashing immediately (at least for my use-case).

**Comment by tagomoris** [comment]
Are there any expected behavior or design for all logging drivers? Or is it expected only for fluentd driver?
In my use case, current behavior is correct because we can get known about Fluentd process's unexpected shutdown before launching containers.

But I'll follow the default (expected) behavior for all drivers if it exists.

**Comment by thaJeztah** [comment]
@tagomoris good question; I'm not able to answer that (but I cc'd you, because you would probably be interested in following this issue);

ping @lk4d4 any thoughts on this?

**Comment by LK4D4** [comment]
I think daemon shouldn't start without logging driver available. Also it shouldn't use fallbacks, because it's unexpected and weird for almost all usecases.

**Comment by newhook** [comment]
I strongly believe the daemon should start without logging. I don't want our services not to startup because logging is temporarily unavilable. Furthermore, the current behavior is inconsistent because the service with continue to run if logging is unavailable, but will refuse to start in the same circumstance.

**Comment by tagomoris** [comment]
@newhook As you said, logging service may fail sometimes, but we can't stop containers for that accidental logging service failure, because that stop occurs whole cluster crash if all containers depends on just one logging service host (even if logging service daemon will be back after several second downtime). So I believe that container shouldn't stop when logging service stops.

On the other hand, if we want to configure logging with one logging service host in one container host, and if we forgot to launch logging host container, we can't find it forever if container starts with logging service outage. We can find the logging service outage and fix it if container doesn't start without logging service.

Any thoughts?

**Comment by newhook** [comment]
The problem is that the absence or presence of the logging service at container start means nothing. It may be there and then immediately disappear, or it may not be there and immediately appear.

**Comment by tagomoris** [comment]
1. "It may be there and then immediately disappear": it's accident. (we can't avoid it)
2. "it may no be there": it's may be configuration mistake, or accident. we can avoid configuration mistake by determine not to launch containers.

**Comment by newhook** [comment]
For my production environments I would never want a temporary network failure to stop my containers from launching. So at the very least I would want to control whether the service launches in the event that the logging service is unreachable.

**Comment by thaJeztah** [comment]
Wondering, are restart policies already active if the container fails to start? In that case, `--restart on-failure` could be a solution; i.e. try to (re)start the container if the logging service is not available. Just thinking out loud here :-)

**Comment by newhook** [comment]
I don't use restart policies, the containers themselves are started/stopped by upstart or systemd.

**Comment by jfindley** [comment]
This also means we can't run fluentd as a container, as there's a chicken-and-egg problem - the log forwarding container can't start because the log forwarding container isn't running.

By all means throw some warning messages at syslog, or have some other way of saying "something is wrong" - but to work at all because the logging endpoint isn't currently present is a nuisance.

**Comment by dpgeekzero** [comment]
I don't see the chicken and egg problem.  The logging driver config is per
container, so not all containers have to go to the same driver.

**Comment by newhook** [comment]
Yeah, I run fluent in a container with no issues.

**Comment by newhook** [comment]
Woops, didn't mean to close the issue. I run fluent in a container with no issues. The logging driver for the fluent container is obviously not fluent itselt.

**Comment by jfindley** [comment]
You can indeed specify logging driver per container, but you can also specify it globally.  What I'm doing now is running each container with --logdriver=fluentd -log-opt ... -- but this adds overhead and nuisance to container management, as it needs to be specified each time.

It also means that if the fluentd container dies, so do all the others.

What I'd like to be able to do is specify it globally at startup, and have docker buffer logs until fluentd becomes available.

**Comment by newhook** [comment]
> It also means that if the fluentd container dies, so do all the others.

No it means you cannot start new containers. Already running containers logging to fluent don't terminate if fluent dies.

**Comment by jnummelin** [comment]
I agree with others saying that my services are more important than logging. Therefore I would really like to see the containers starting even if the fluentd is not available. The current behaviour also means that if (and when) I run my fluentd in a container, I have to somehow control the starting order of my containers in case the host reboots. If you do not want to change the existing behaviour maybe introduce a new log-opt flag for ignoring connect errors at container startup.

**Comment by cpuguy83** [comment]
If someone wants to make a PR that'd be great.

**Comment by gtmtech** [comment]
+1 Really glad to see this is being fixed.

**Comment by thaJeztah** [comment]
Fix for this has been merged, and will be in the 1.11 release (it just didn't make the deadline for 1.10); you can try the "experimental" builds though, who should have this (on the next build).

**Comment by MrMMorris** [comment]
@thaJeztah now THIS is what I was missing when I ran into this problem: https://github.com/docker/docker/issues/21781#issuecomment-206113979

Yay solutions!
